feat(releases): add support for COTS clearing

alexbrdn opened the pull request:
- add component type to clearing status tab of project details view.
- add new fields (cotsResponsible, clearingDeadline, containsOSS) to COTSDetails for supporting COTS clearing
- reorganize Commercial Details into two tables

closes sw360/sw360portal#503

mcjaeger commented:
@adrian-evo if you would have time, mind a retest?

adrian-evo commented:
Noticed two issues:

- At creation time of a new release for a COTS component, the boolean fields are not saved. These are saved only if edited afterwards and updated.
- The Commercial Details are not listed under moderation request for COTS releases

maierthomas commented on backend/src-common/src/test/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandlerTest.java:
```diff
@@ -279,4 +277,40 @@ public void testSanityCheckSucceeds() throws Exception {
         RequestStatus status2 = handler.updateProject(project2, user1);
         assertThat(status2, is(RequestStatus.SUCCESS));
     }
-}
```
New line is missing here

maierthomas commented on frontend/sw360-portlet/src/main/webapp/html/components/includes/releases/editCommercialDetails.jspf:
```diff
@@ -65,6 +87,8 @@
                         <core_rt:if test="${release.cotsDetails.ossContractSigned == 'true'}"> checked="checked" </core_rt:if>
                 /> OSS contract signed</label>
```
This is only a small one. The first letter should be in uppercase like in all the other labels.
OSS Contract Signed

alexbrdn commented on backend/src-common/src/test/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandlerTest.java:
```diff
@@ -279,4 +277,40 @@ public void testSanityCheckSucceeds() throws Exception {
         RequestStatus status2 = handler.updateProject(project2, user1);
         assertThat(status2, is(RequestStatus.SUCCESS));
     }
-}
```
what line break?

alexbrdn commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:
```diff
@@ -478,6 +468,80 @@ public RequestStatus updateProjectVulnerabilityRating(ProjectVulnerabilityRating
+    }
+
+    SetMultimap<String, ProjectWithReleaseRelationTuple> releaseIdToProjects(Project project, User user) throws SW360Exception {
+        Set<String> visitedProjectIds = new HashSet<>();
```
The graph of linked projects is not guaranteed to be acyclic.

alexbrdn commented on frontend/sw360-portlet/src/main/webapp/html/components/includes/releases/editCommercialDetails.jspf:
```diff
@@ -32,24 +28,50 @@
+            <input id="cotsClearingDeadline"
+                   class="datepicker"
+                   name="<portlet:namespace/><%=Release._Fields.COTS_DETAILS%><%=COTSDetails._Fields.CLEARING_DEADLINE%>"
                    type="text"
```
javascript datepicker takes care of that

alexbrdn commented on frontend/sw360-portlet/src/main/webapp/html/components/includes/releases/editCommercialDetails.jspf:
```diff
@@ -65,6 +87,8 @@
                         <core_rt:if test="${release.cotsDetails.ossContractSigned == 'true'}"> checked="checked" </core_rt:if>
```
No, in this case it's not. I'll clean this up.

alexbrdn commented:
@maierthomas, @adrian-evo, I beleive I have fixed the problems. Please re-review.
Also fixed the accepting of a moderation request for deletion, which used to throw an NPE.

adrian-evo commented:
After creating a COTS type of release successfully, I cannot create a project with linked COTS type of release. The exception is:

 2017-10-06 10:37:13 ERROR Sw360ThriftServlet:44 - uncaught
 org.ektorp.DbAccessException: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "cotsResponsible" (class org.eclipse.sw360.datahandler.thrift.components.COTSDetails), not marked as ignorable (13 known properties: "usedLicense", "licenseClearingReportURL", "ossInformationURLIsSet", "containsOSSIsSet", "usageRightAvailableIsSet", "usageRightAvailable", "containsOSS", "ossInformationURL", "licenseClearingReportURLIsSet" [truncated]])
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.eclipse.sw360.datahandler.thrift.components.Release["cotsDetails"]->org.eclipse.sw360.datahandler.thrift.components.COTSDetails["cotsResponsible"])
	at org.ektorp.util.Exceptions.propagate(Exceptions.java:19)

alexbrdn commented:
@adrian-evo, this looks like a case of a partial deployment to me. Make sure there are no `.war` files in the tomcat's deploy folder left. (If there are, `touch <something>.war`)
I've just retested this and had no problems at all.

alexbrdn commented:
fixed the build

mcjaeger commented:
@adrian-evo if you would have time, mind a test?

adrian-evo commented:
Moderation works, but now COTS responsible, Clearing Deadline and Source Code Available are not saved.

alexbrdn commented:
I cannot reproduce any such problem. I've asked @maierthomas to double-check.

adrian-evo commented:
Tried again with a new clean deployment and it works as expected.
